File: src/lib/ratings.ts
import type { Phone, RatingCategory, UserVote } from "@/types/phone";

export const RATING_KEY = "cepkiyas.votes.v1";

export const ratingCategories: { id: RatingCategory; label: string }[] = [
  { id: "overall", label: "Genel" },
  { id: "camera", label: "Kamera" },
  { id: "performance", label: "Performans" },
  { id: "battery", label: "Batarya" },
  { id: "display", label: "Ekran" },
  { id: "design", label: "Tasarım" },
];

export function emptyVote(): UserVote {
  return {
    overall: 0,
    camera: 0,
    performance: 0,
    battery: 0,
    display: 0,
    design: 0,
  };
}

export function readVotes(): Record<string, UserVote> {
  if (typeof window === "undefined") return {};
  try {
    const raw = window.localStorage.getItem(RATING_KEY);
    return raw ? (JSON.parse(raw) as Record<string, UserVote>) : {};
  } catch {
    return {};
  }
}

export function writeVotes(votes: Record<string, UserVote>) {
  window.localStorage.setItem(RATING_KEY, JSON.stringify(votes));
}

export function mergedUserScore(phone: Phone, vote?: UserVote) {
  const seedTotal = phone.seedRatings.average * phone.seedRatings.count;
  const extra = vote?.overall ? vote.overall : 0;
  const extraCount = vote?.overall ? 1 : 0;
  const count = phone.seedRatings.count + extraCount;
  const average = (seedTotal + extra) / count;

  const breakdown = {
    camera: blend(phone.seedRatings.camera, vote?.camera),
    performance: blend(phone.seedRatings.performance, vote?.performance),
    battery: blend(phone.seedRatings.battery, vote?.battery),
    display: blend(phone.seedRatings.display, vote?.display),
    design: blend(phone.seedRatings.design, vote?.design),
  };

  return {
    average: Math.round(average * 10) / 10,
    count,
    breakdown,
    mine: vote,
  };
}

function blend(seed: number, mine?: number) {
  if (!mine) return seed;
  return Math.round(((seed * 12 + mine) / 13) * 10) / 10;
}

File: mobile/lib/logic/ratings.dart
import '../models/phone.dart';

const ratingCategories = [
  ('overall', 'Genel'),
  ('camera', 'Kamera'),
  ('performance', 'Performans'),
  ('battery', 'Batarya'),
  ('display', 'Ekran'),
  ('design', 'Tasarım'),
];

class MergedScore {
  const MergedScore({
    required this.average,
    required this.count,
    required this.breakdown,
    this.mine,
  });

  final double average;
  final int count;
  final Map<String, double> breakdown;
  final UserVote? mine;
}

MergedScore mergedUserScore(Phone phone, UserVote? vote) {
  final extra = vote != null && vote.overall > 0 ? vote.overall : 0;
  final extraCount = extra > 0 ? 1 : 0;
  final count = phone.seedRatings.count + extraCount;
  final average = (phone.seedRatings.average * phone.seedRatings.count + extra) / count;
  return MergedScore(
    average: (average * 10).round() / 10,
    count: count,
    breakdown: {
      'camera': _blend(phone.seedRatings.camera, vote?.camera),
      'performance': _blend(phone.seedRatings.performance, vote?.performance),
      'battery': _blend(phone.seedRatings.battery, vote?.battery),
      'display': _blend(phone.seedRatings.display, vote?.display),
      'design': _blend(phone.seedRatings.design, vote?.design),
    },
    mine: vote,
  );
}

double _blend(double seed, int? mine) {
  if (mine == null || mine == 0) return seed;
  return ((seed * 12 + mine) / 13 * 10).round() / 10;
}
